partition-alloc/src/allocator_shim.cc

```diff
@@ -0,0 +1,221 @@
+  // we see unexpected PKRU values.
+  default: {
+    return 0;
+  }
+  }
+}
+#endif // LIBIA2_INSECURE
+
+using namespace partition_alloc::internal;
+using partition_alloc::PartitionOptions;
+using partition_alloc::PartitionRoot;
+using partition_alloc::internal::PartitionAllocator;
+
+class SimpleScopedSpinLocker {
+public:
+  explicit SimpleScopedSpinLocker(std::atomic<bool> &lock) : lock_(lock) {
+    bool expected = false;
+    while (!lock_.compare_exchange_weak(
+        expected, true, std::memory_order_acquire, std::memory_order_relaxed)) {
+      expected = false;
+    }
+  }
+
+  ~SimpleScopedSpinLocker() { lock_.store(false, std::memory_order_release); }
+
+private:
+  std::atomic<bool> &lock_;
+};
+
+// This is the LeakySingleton class in
+// allocator_shim_default_dispatch_to_partition_alloc.cc, but it's renamed here
+// since we don't need it to be generic over the type of partition root.
+class SingletonPartition {
+public:
+  PartitionRoot<ThreadSafe> *Get() {
+    auto *instance = instance_.load(std::memory_order_acquire);
+    if (instance) {
+      return instance->root();
+    }
+    return GetSlowPath();
+  }
+
+private:
+  PartitionRoot<ThreadSafe> *GetSlowPath();
+  std::atomic<PartitionAllocator<ThreadSafe> *> instance_;
+  alignas(PartitionAllocator<ThreadSafe>) uint8_t
+      instance_buffer_[sizeof(PartitionAllocator<ThreadSafe>)] = {0};
+  std::atomic<bool> initialization_lock_;
+};
+
+static PartitionAllocator<ThreadSafe> *NewPartition(void *buffer) {
+  auto *new_heap = new (buffer) PartitionAllocator<ThreadSafe>();
+  new_heap->init({
+      PartitionOptions::AlignedAlloc::kDisallowed,
+      PartitionOptions::ThreadCache::kDisabled,
+      PartitionOptions::Quarantine::kDisallowed,
+      PartitionOptions::Cookie::kDisallowed,
+      PartitionOptions::BackupRefPtr::kDisabled,
+      PartitionOptions::BackupRefPtrZapping::kDisabled,
+      PartitionOptions::UseConfigurablePool::kNo,
+  });
+  return new_heap;
+}
+
+PartitionRoot<ThreadSafe> *SingletonPartition::GetSlowPath() {
+  SimpleScopedSpinLocker scoped_lock{initialization_lock_};
+
+  PartitionAllocator<ThreadSafe> *instance =
+      instance_.load(std::memory_order_relaxed);
+
+  if (instance) {
+    return instance->root();
+  }
+
+  instance = ::NewPartition(reinterpret_cast<void *>(instance_buffer_));
+  instance_.store(instance, std::memory_order_release);
+
+  return instance->root();
+}
+
+SingletonPartition g_partitions[16];
+
+void *ShimMalloc(size_t bytes) {
+  size_t pkey = ::ia2_get_pkey();
+  return ShimMallocWithPkey(bytes, pkey);
+}
+
+void *ShimMallocWithPkey(size_t bytes, size_t pkey) {
+  if (pkey == 0) {
+    return malloc(bytes);
+  } else {
+    PartitionRoot<ThreadSafe> *root = g_partitions[pkey].Get();
+    return root->Alloc(bytes, nullptr);
+  }
+}
+
+void ShimFree(void *ptr) {
+  size_t pkey = ::ia2_get_pkey();
+  ShimFreeWithPkey(ptr, pkey);
+}
+
+void ShimFreeWithPkey(void *ptr, size_t pkey) {
```

Comment:
Can this (or `ShimReallocWithPkey` or `ShimCallocWithPkey`) be called maliciously with another compartment's pkey?

Reply:
I don't see why not. Looking at `SHIM_ALWAYS_EXPORT`, I think this lib is meant to be compiled with `-fvisibility=hidden`. Even then, would that prevent someone from jumping to the address just like the exported methods do?

@ayrtonm?

Reply:
Yeah this can currently be called with another compartment's pkey and fixing it would be part of #165. We can prevent this kind of attack with the second idea in [this comment](https://github.com/immunant/IA2-Phase2/issues/165#issue-1370194839) where we put the partition roots on separate pages with each compartment's pkey.